contracts/src/TeleporterTokenSource.sol

```diff
@@ -197,65 +375,69 @@ abstract contract TeleporterTokenSource is
+     * - {input.destinationBridgeAddress} is this contract is this chain is the final destination.
      */
     function _receiveTeleporterMessage(
         bytes32 sourceBlockchainID,
         address originSenderAddress,
         bytes memory message
     ) internal override {
         BridgeMessage memory bridgeMessage = abi.decode(message, (BridgeMessage));
-
-        // Check that bridge instance returning has sufficient amount in balance
-        uint256 senderBalance = bridgedBalances[sourceBlockchainID][originSenderAddress];
-        require(
-            senderBalance >= bridgeMessage.amount,
-            "TeleporterTokenSource: insufficient bridge balance"
-        );
-
-        // Decrement the bridge balance by the unwrap amount
-        bridgedBalances[sourceBlockchainID][originSenderAddress] =
-            senderBalance - bridgeMessage.amount;
-
         if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_SEND) {
             SingleHopSendMessage memory payload =
                 abi.decode(bridgeMessage.payload, (SingleHopSendMessage));
-            _withdraw(payload.recipient, bridgeMessage.amount);
+
+            uint256 sourceAmount =
+                _processReceivedTransfer(sourceBlockchainID, originSenderAddress, payload.amount);
+
+            // Send the tokens to the recipient.
+            _withdraw(payload.recipient, sourceAmount);
             return;
         } else if (bridgeMessage.messageType == BridgeMessageType.SINGLE_HOP_CALL) {
             SingleHopCallMessage memory payload =
                 abi.decode(bridgeMessage.payload, (SingleHopCallMessage));
 
-            // Verify that the payload's source blockchain ID
-            // matches the source blockchain ID passed from Teleporter.
-            // Prevents a destination bridge from accessing tokens attributed
-            // to another destination bridge instance.
+            uint256 sourceAmount =
+                _processReceivedTransfer(sourceBlockchainID, originSenderAddress, payload.amount);
+
+            // Verify that the payload's source blockchain ID matches the source blockchain ID passed from Teleporter.
+            // Prevents a destination bridge from accessing tokens attributed to another destination bridge instance.
             require(
                 payload.sourceBlockchainID == sourceBlockchainID,
                 "TeleporterTokenSource: mismatched source blockchain ID"
             );
-            _handleSendAndCall(payload, bridgeMessage.amount);
+
+            _handleSendAndCall(payload, sourceAmount);
             return;
         } else if (bridgeMessage.messageType == BridgeMessageType.MULTI_HOP_SEND) {
             MultiHopSendMessage memory payload =
                 abi.decode(bridgeMessage.payload, (MultiHopSendMessage));
+
+            uint256 sourceAmount =
+                _processReceivedTransfer(sourceBlockchainID, originSenderAddress, payload.amount);
+
```

Comment:
Calling out that the `payload.secondaryFee` amount should be scaled back to the source denomination here prior to calling `_send` or `_sendAndCall`. I think this will be done in #109

Reply:
Yeah, it's not done yet, I was going to wait until the logic from the two PRs are integrated, since this PR changes how all the scaling is done.

Reply:
Ah sorry, just put the same comment on the other PR before seeing this comment

Reply:
Sounds good though